core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -68,37 +60,26 @@ object RequestChannel extends Logging {
     @volatile var apiRemoteCompleteTimeNanos = -1L
     @volatile var recordNetworkThreadTimeCallback: Option[Long => Unit] = None
 
-    val header: RequestHeader = try {
-      RequestHeader.parse(buffer)
-    } catch {
-      case ex: Throwable =>
-        throw new InvalidRequestException(s"Error parsing request header. Our best guess of the apiKey is: ${buffer.getShort(0)}", ex)
-    }
+    val session = Session(context.principal, context.clientAddress)
+    val request: RequestAndSize = context.parseRequest(buffer)
 
-    val bodyAndSize: RequestAndSize =
-      try {
-        // For unsupported version of ApiVersionsRequest, create a dummy request to enable an error response to be returned later
-        if (header.apiKey == ApiKeys.API_VERSIONS.id && !Protocol.apiVersionSupported(header.apiKey, header.apiVersion)) {
-          new RequestAndSize(new ApiVersionsRequest.Builder().build(), 0)
-        }
-        else
-          AbstractRequest.getRequest(header.apiKey, header.apiVersion, buffer)
-      } catch {
-        case ex: Throwable =>
-          throw new InvalidRequestException(s"Error getting request for apiKey: ${header.apiKey} and apiVersion: ${header.apiVersion}", ex)
-      }
+    def header: RequestHeader = context.header
+    def connectionId: String = context.connectionId
+    def listenerName: ListenerName = context.listenerName
+    def securityProtocol: SecurityProtocol = context.securityProtocol
```

Comment:
I think I'll keep the forwarder for `header` since we have one for the request body already and there are quite a lot of usages, but I can remove the others.